src/node/txreconciliation.cpp

```diff
@@ -357,79 +357,80 @@ class TxReconciliationTracker::Impl
+        // - for outbounds we have a fixed number of flood destinations.
+        // - for inbounds we use a fraction of all inbound peers supporting tx relay.
+        size_t outbounds_target_size = 0;
+        if (OUTBOUND_FANOUT_DESTINATIONS > outbounds_fanout_tx_relay) {
+            outbounds_target_size = OUTBOUND_FANOUT_DESTINATIONS - outbounds_fanout_tx_relay;
+        }
 
+        // Since we use the fraction for inbound peers, we first need to compute the total number of inbound targets.
+        const double inbound_targets = (inbounds_fanout_tx_relay + m_inbounds_count) * INBOUND_FANOUT_DESTINATIONS_FRACTION;
+        double n = std::max(inbound_targets - inbounds_fanout_tx_relay, 0.0);
+
+        // Being this a fraction, we need to round it either up or down. We do this deterministically at random based on the
+        // transaction we are picking the peers for.
+        CSipHasher deterministic_randomizer_in{m_deterministic_randomizer};
+        deterministic_randomizer_in.Write(wtxid.ToUint256());
+        CSipHasher deterministic_randomizer_out{deterministic_randomizer_in};
+        const size_t inbounds_target_size = ((deterministic_randomizer_in.Finalize() & 0xFFFFFFFF) + uint64_t(n * 0x100000000)) >> 32;
+
+        // Pick all reconciliation registered peers and assign them a deterministically random value based on their peer id
+        // Also, split peers in inbounds/outbounds
+        std::vector<std::pair<uint64_t, NodeId>> sorted_inbounds, sorted_outbounds;
+        sorted_inbounds.reserve(m_inbounds_count);
+        Assume(m_states.size() >= m_inbounds_count);
+        sorted_outbounds.reserve(m_states.size() - m_inbounds_count);
         for (const auto& [node_id, op_peer_state]: m_states) {
             const auto peer_state = std::get_if<TxReconciliationState>(&op_peer_state);
-            if (peer_state && peer_state->m_we_initiate == we_initiate) {
-                uint64_t hash_key = CSipHasher(deterministic_randomizer).Write(node_id).Finalize();
-                best_peers.emplace_back(hash_key, node_id);
+            if (peer_state) {
+                if (peer_state->m_we_initiate) {
+                    uint64_t hash_key = CSipHasher(deterministic_randomizer_out).Write(node_id).Finalize();
+                    sorted_outbounds.emplace_back(hash_key, node_id);
+                } else {
+                    uint64_t hash_key = CSipHasher(deterministic_randomizer_in).Write(node_id).Finalize();
+                    sorted_inbounds.emplace_back(hash_key, node_id);
+                }
+
             }
         }
 
-        std::sort(best_peers.begin(), best_peers.end());
-
-        auto it = best_peers.begin();
-        for (size_t i = 0; i < targets_size && it != best_peers.end(); ++i, ++it) {
-            if (it->second == peer_id) return true;
+        // Sort the peers based on their assigned random value, extract the node_ids and trim the collections to size
```

Comment:
I think this is slightly harder to follow, but I like that the collections are only sorted up to some point, so I'll take it. Instead of using the `std::min`, I've added two `Assume` when defining `{outbounds, inbounds}_target_size` to ensure that the targets are never bigger than the sets the pull from, since this should never be the case

Reply:
Covered in [4b4d99f](https://github.com/bitcoin/bitcoin/pull/30116/commits/4b4d99fb2df2c60a2214487cec627bc560f50f53)